### ISSUE TITLE
CHECK_THROWS silently works when CHECK_THROWS_WITH was intented.

### ISSUE THREAD

**Issue by J-Richter** [doctest/doctest#737]
## Description

When checking for exception CHECK_THROWS or CHECK_THROWS_WITH can be used.
Both accept the same syntax:  ( expression, string ).
When CHECK_THROWS is wrongly used instead of CHECK_THROWS_WITH, it works without problems, but the error text is not checked.

### Steps to reproduce

CHECK_THROWS( throw std::runtime_error( "hello" ), "world" );

works. But CHECK_THROWS_WITH was intented. Which would fail.

**Comment by J-Richter** [comment]
Suggested solution is to forbid calling CHECK_THROWS with more than one argument.

**Comment by cdeln** [comment]
Reason is that `CHECK_THROWS` is currently defined as `#define CHECK_THROWS(...) DOCTEST_CHECK_THROWS(__VA_ARGS__)`
and `DOCTEST_CHECK_THROWS` is defined as
`#define DOCTEST_CHECK_THROWS(...) DOCTEST_ASSERT_THROWS_WITH((__VA_ARGS__), #__VA_ARGS__, DT_CHECK_THROWS, "")`
Not sure why `DOCTEST_CHECK_THROWS` is variadic.

**Comment by mitchgrout** [comment]
> Suggested solution is to forbid calling CHECK_THROWS with more than one argument.

Seems the `_WITH`, `_WITH_AS`, and `_AS` family of checks just take an `expr` here. We could convert, but it could be considered an API breakage. See the next section for context.

> Not sure why `DOCTEST_CHECK_THROWS` is variadic.

tl;dr preprocessor screws up commas unless they're in `(parens)`, so variadic args are required to deal with things like `CHECK_NOTHROW( fn<A, B>() )`

Suppose you have the following macros:

```cpp
#define A(x, ...) x
#define B(x) x
#define C(...) __VA_ARGS__
```

You now want to invoke them with an expression that involves a comma, and that comma is **not** inside parens. Let's say, `x[y, z]`. So we would have:

```cpp
A( x[y, z] )// x[y
B( x[y, z] ) // error: macro 'B' passed 2 arguments, but takes just 1
C( x[y, z] ) // x[y, z]
```

The preprocessor does not parse `,` in the same way as the C++ language; it will eagerly split args by `,` (unless inside `(parens)`). The `C` macro is the only one that respects the `,` (it still splits on it, but glues all the args back together correctly). So, the general rule is, if you need to capture an expression, use `__VA_ARGS__`.

Internally, `DOCTEST_CHECK_NOTHROW` resolves to `DOCTEST_ASSERT_THROWS_WITH((__VA_ARGS__), ...)`. The `(__VA_ARGS__)` here ensures that the preprocessor doesn't try to split up on commas again. But this unfortunately forms a [comma-expression](https://en.cppreference.com/w/cpp/language/operator_other.html#Built-in_comma_operator), which is why your example still compiles; you have:

```cpp
CHECK_NOTHROW( throw std::runtime_error("hello"), "world" );
// Expands to (roughly):
try {
  static_cast<void>( (throw std::runtime_error("hello"), "world") );
  // Semantically will:
  // - Throw the exception, then
  // - Return the result "world" (obviously we never _actually_ do this since the first step is noreturn
}
```

If we swap from `__VA_ARGS__` to just a single `expr`, we probably won't see many breakages, but it's a possibility. At least, it won't be silent: the compiler will give a nice noisy warning about too many arguments to a macro.
